### PR TITLE
[6188] Add presence start-end function.  

### PR DESCRIPTION
**Pull request by mars22**
[6188] Add presence start-end function.
1.new screen added to app
2.backend endpoint for this screen

https://bugzilla.stxnext.pl/show_bug.cgi?id=6188

### CODE REVIEW

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -83,6 +84,32 @@ def group_by_weekday(items):
     return result
 
 
+def start_end_group_by_weekday(items):
+    """
+    Groups presence entries by weekday. Each weekday contains to list.
+    First list contains all starts for given day
+    Secound list contains all ends for given day
+    """
+    result = [[[], []], [[], []], [[], []], [[], []], [[], []], [[], []], [[],
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: That's one nested hell of lists :fearful:. Perhaps a dict would be more suited for the job?
If I understood correctly: 
```
[
    1, [[start, start, start], [end, end, end]],
    2, [[start, start, start], [end, end, end]],
    ...
]
```
could be easily swapped for:
```
{
    1: {'start': [start, start, start], 'end': [end, end, end]},
    2: {'start': [start, start, start], 'end': [end, end, end]},
    ...
}
```
Personally I find the latter way more readable and easy to understand than the nested lists. What do you think?

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -53,8 +54,22 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_api_star_end(self):
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Typo `star_end => start_end`.

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -53,8 +54,22 @@ def test_api_users(self):
...
+        """
+        Test mean start and end presence of user per day
+        """
+        resp = self.client.get('/api/v1/presence_start_end_per_weekday/11')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 7)
+        self.assertEqual(data[3][0], u'Thu')
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: I'd prefer to see the whole result compared than certain part of it. Easier to see the structure in tests. This applies to other tests as well.

Comment (mars22) [inline src/presence_analyzer/tests.py]: You mean something like this
```
self.assertEqual(data[3], [u'Thu', u'09:53:22', u'16:16:26'])
```

File: src/presence_analyzer/utils.py (hunk 1 above)
Comment (mars22) [inline src/presence_analyzer/utils.py]: I've got absolutly right :)

File: src/presence_analyzer/tests.py (hunk 3 above)
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: More like, you don't have to check the length then.
```
self.assertEqual(
    data,
    [
         ...
         [u'Thu', u'09:53:22', u'16:16:26'],
         [u'Fri', u'09:53:22', u'16:16:26'],
         ...
    ]
)

File: src/presence_analyzer/static/presence_start_end.html [hunk 4]
```diff
@@ -97,31 +102,33 @@
...
+                        
+                        $.getJSON("/api/v1/presence_start_end_per_weekday/"+selected_user, function(result) {
+                            var data = new google.visualization.DataTable();
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            var convertedResults =[];
+                            $.each(result,function(i,item){
+                                convertedResults.push([item[0],convertTimeToDate(item[1]),convertTimeToDate(item[2])]);    
+
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: You don't need this blank line.

File: src/presence_analyzer/static/presence_start_end.html [hunk 5]
```diff
@@ -97,31 +102,33 @@
...
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            var convertedResults =[];
+                            $.each(result,function(i,item){
+                                convertedResults.push([item[0],convertTimeToDate(item[1]),convertTimeToDate(item[2])]);    
+                                
+                            });
+                            
+                            data.addRows(convertedResults);
+                            var options = {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: How to declare few variables in js:
```
var a = 123123,
    b = 2313,
    c = 342939;
````

File: src/presence_analyzer/static/presence_start_end.html [hunk 6]
```diff
@@ -97,31 +102,33 @@
...
-                            ["Thu", new Date(1, 1, 1, 7, 26, 32), new Date(1, 1, 1, 12, 40, 45)],
-                            ["Fri", new Date(1, 1, 1, 6, 53, 22), new Date(1, 1, 1, 12, 52, 09)]
-                        ] 
+                        
+                        $.getJSON("/api/v1/presence_start_end_per_weekday/"+selected_user, function(result) {
+                            var data = new google.visualization.DataTable();
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            var convertedResults =[];
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: ```var convertedResults = [];```

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -53,8 +54,28 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_api_start_end(self):
+        """
+        Test mean start and end presence of user per day
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Missing full stop on the end of docstring.

File: src/presence_analyzer/utils.py [hunk 8]
```diff
@@ -83,6 +83,55 @@ def group_by_weekday(items):
...
+        end = items[date]['end']
+        curr = result.setdefault(date.weekday(), {'start': [], 'end': []})
+        curr['start'].append(seconds_since_midnight(start))
+        curr['end'].append(seconds_since_midnight(end))
+    return result
+
+
+def str_to_time(str_time):
+    """
+    convert string rep. of time to time
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: ```Convert string rep. of time to time.```

File: src/presence_analyzer/utils.py [hunk 9]
```diff
@@ -83,6 +83,55 @@ def group_by_weekday(items):
...
+
+
+def mean_start_end_by_weekday(grouped):
+    """
+    Calculate mean start/end time by day
+    """
+
+    result = [
+        [
+            str_to_time(
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]:
```
str_to_time(mean(item['start'])),
str_to_time(mean(item['end'])),
```

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -53,8 +54,28 @@ def test_api_users(self):
...
+        """
+        resp = self.client.get('/api/v1/presence_start_end_per_weekday/11')
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        data = json.loads(resp.data)
+        self.assertEqual(
+            data,
+            [
+                [u'Mon', u'09:12:14', u'15:54:17'],
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ```from __future__ import unicode_literals```
Then you don't need to use ```u```

File: src/presence_analyzer/static/presence_start_end.html [hunk 11]
```diff
@@ -90,6 +90,11 @@
                     dropdown.show();
                     loading.hide();
                 });
+
+                function convertTimeToDate(time){
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: ```function convertTimeToDate(time) {```

File: src/presence_analyzer/static/presence_start_end.html [hunk 12]
```diff
@@ -97,31 +102,33 @@
...
-                            ["Fri", new Date(1, 1, 1, 6, 53, 22), new Date(1, 1, 1, 12, 52, 09)]
-                        ] 
+                        
+                        $.getJSON("/api/v1/presence_start_end_per_weekday/"+selected_user, function(result) {
+                            var data = new google.visualization.DataTable();
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            var convertedResults =[];
+                            $.each(result,function(i,item){
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: $.each(result,function(i,item) {

File: src/presence_analyzer/static/presence_start_end.html [hunk 13]
```diff
@@ -90,38 +90,44 @@
...
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            $.each(result, function(i,item) {
+                                var day   = item[0],
+                                    start = convertTimeToDate(item[1]),
+                                    end   = convertTimeToDate(item[2]),
+                                    
+                                convertedResults.push([day, start, end]);    
+
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: Delete this line.

File: src/presence_analyzer/tests.py [hunk 14]
```diff
@@ -17,6 +18,7 @@
 
 # pylint: disable=maybe-no-member, too-many-public-methods
 class PresenceAnalyzerViewsTestCase(unittest.TestCase):
+
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Delete this line.

File: src/presence_analyzer/tests.py [hunk 15]
```diff
@@ -51,10 +53,30 @@ def test_api_users(self):
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
         self.assertEqual(len(data), 2)
-        self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
+        self.assertDictEqual(data[0], {'user_id': 10, 'name': 'User 10'})
+
+    def test_api_start_end(self):
+        """
+        Test mean start and end presence of user per day
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Full stop.

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -86,6 +108,44 @@ def test_get_data(self):
...
         )
 
+    def test_start_end_grouped_by_weekday(self):
+        """
+        Test grouped start end presence by weekday.
+        """
+        data = utils.get_data()
+        user_data = data[11]
+        grouped_by_weekday = utils.start_end_group_by_weekday(user_data)
+        print grouped_by_weekday
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: You need this?

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -86,6 +108,44 @@ def test_get_data(self):
             datetime.time(9, 39, 5)
         )
 
+    def test_start_end_grouped_by_weekday(self):
+        """
+        Test grouped start end presence by weekday.
+        """
+        data = utils.get_data()
+        user_data = data[11]
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: You don't have to declare variable if you are using it only once.

File: src/presence_analyzer/tests.py [hunk 18]
```diff
@@ -51,10 +53,30 @@ def test_api_users(self):
...
         self.assertEqual(len(data), 2)
-        self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
+        self.assertDictEqual(data[0], {'user_id': 10, 'name': 'User 10'})
+
+    def test_api_start_end(self):
+        """
+        Test mean start and end presence of user per day
+        """
+        resp = self.client.get('/api/v1/presence_start_end_per_weekday/11')
+        self.assertEqual(resp.status_code, 200)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Also you should check if not existing user return 404.

File: src/presence_analyzer/tests.py [hunk 19]
```diff
@@ -2,6 +2,7 @@
 """
 Presence analyzer unit tests.
 """
+from __future__ import unicode_literals
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Empty line after this import.

File: src/presence_analyzer/static/presence_start_end.html [hunk 20]
```diff
@@ -90,38 +90,44 @@
...
+                            chart_div.show();
+                            loading.hide();
+                            chart.draw(data, options);
+                        });
 
-                        chart_div.show();
-                        loading.hide();
-                        var chart = new google.visualization.Timeline(chart_div[0]);
-                        chart.draw(data, options);
+
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: To many blank lines.

File: src/presence_analyzer/static/presence_start_end.html [hunk 21]
```diff
@@ -90,39 +90,39 @@
...
+                            var data = new google.visualization.DataTable(),
+                                convertedResults = [],
+                                options = { hAxis: {title: 'Weekday'}},
+                                formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'}),
+                                chart = new google.visualization.Timeline(chart_div[0]);
+                            data.addColumn('string', 'Weekday');
+                            data.addColumn({ type: 'datetime', id: 'Start' });
+                            data.addColumn({ type: 'datetime', id: 'End' });
+                            $.each(result, function(i,item) {
+                                var day   = item[0],
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: to many spaces

File: src/presence_analyzer/static/presence_start_end.html [hunk 22]
```diff
@@ -90,39 +90,39 @@
...
+                                    start = convertTimeToDate(item[1]),
+                                    end   = convertTimeToDate(item[2]),
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: Line should end with ```;```

File: src/presence_analyzer/tests.py [hunk 23]
```diff
@@ -51,7 +53,36 @@ def test_api_users(self):
...
+            resp = self.client.get("/api/v1/presence_start_end_per_weekday/{0}"
+                                   .format(user_id))
+            self.assertEqual(resp.status_code, 404)
+
+    def test_api_start_end(self):
+        """
+        Test mean start and end presence of user per day.
+        """
+        resp = self.client.get('/api/v1/presence_start_end_per_weekday/11')
+        self.assertEqual(resp.status_code, 200)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: More cases

**Comment by dpolcyn** [comment]
Please squash commits

File: src/presence_analyzer/tests.py [hunk 24]
```diff
@@ -51,7 +53,33 @@ def test_api_users(self):
...
         data = json.loads(resp.data)
         self.assertEqual(len(data), 2)
-        self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
+        self.assertDictEqual(data[0], {'user_id': 10, 'name': 'User 10'})
+
+    def test_api_start_end_retun_404(self):
+        """
+        Test mean start and end presence return 404 if can't find user data.
+        """
+        resp = self.client.get("/api/v1/presence_start_end_per_weekday/0")
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: I'm against mixing `'` and `"`. Choose one and use it everywhere, so it's consistent in the whole project.

**Pull request merged**